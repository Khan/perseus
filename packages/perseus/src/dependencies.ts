import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
