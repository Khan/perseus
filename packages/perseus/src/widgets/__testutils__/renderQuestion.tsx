import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
