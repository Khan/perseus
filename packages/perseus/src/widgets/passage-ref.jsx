/* eslint-disable react/sort-comp */
// @flow
import * as i18n from "@khanacademy/wonder-blocks-i18n";
import * as React from "react";
import _ from "underscore";

import * as Changeable from "../mixins/changeable.jsx";
import {removeDenylistProps} from "../mixins/widget-prop-denylist.js";
import PerseusMarkdown from "../perseus-markdown.jsx";

import PassageWidgetExport from "./passage.jsx";

import type {PerseusPassageRefWidgetOptions} from "../perseus-types.js";
import type {
    ChangeFn,
    PerseusScore,
    WidgetExports,
    WidgetProps,
    LinterContextProps,
} from "../types.js";
import type {Reference} from "./passage.jsx";

const EN_DASH = "\u2013";

type UserInput = {|
    ...RenderProps,
    static: ?boolean,
    reviewModeRubric: Rubric,
    linterContext: LinterContextProps,
    isLastUsedWidget: boolean,
    alignment: ?string,
|};

type RenderProps = {|
    passageNumber: PerseusPassageRefWidgetOptions["passageNumber"],
    referenceNumber: PerseusPassageRefWidgetOptions["referenceNumber"],
    summaryText: PerseusPassageRefWidgetOptions["summaryText"],
|};

type Rubric = PerseusPassageRefWidgetOptions;

type Props = WidgetProps<RenderProps, Rubric>;

type DefaultProps = {|
    passageNumber: Props["passageNumber"],
    referenceNumber: Props["referenceNumber"],
    summaryText: Props["summaryText"],
|};

type State = {|
    lineRange: ?[number, number],
    content: ?string,
|};

class PassageRef extends React.Component<Props, State> {
    displayName: string = "PassageRef";
    _isMounted: boolean;
    _throttledUpdateRange: () => void;

    static defaultProps: DefaultProps = {
        passageNumber: 1,
        referenceNumber: 1,
        summaryText: "",
    };

    state: State = {
        lineRange: null,
        content: null,
    };

    shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
        return (
            !_.isEqual(this.props, nextProps) ||
            !_.isEqual(this.state, nextState)
        );
    }

    // TODO(tamarab): getUserInput needs to be updated to only return
    // props input by the user. Currently returns all the widget's props.
    getUserInput: () => UserInput = () => {
        return removeDenylistProps(this.props);
    };

    render(): React.Node {
        const lineRange = this.state.lineRange;
        let lineRangeOutput;
        if (!lineRange) {
            lineRangeOutput = i18n.$_("lines %(lineRange)s", {
                lineRange: `?${EN_DASH}?`,
            });
        } else if (lineRange[0] === lineRange[1]) {
            lineRangeOutput = i18n.$_("line %(lineNumber)s", {
                lineNumber: lineRange[0],
            });
        } else {
            lineRangeOutput = i18n.$_("lines %(lineRange)s", {
                lineRange: lineRange[0] + EN_DASH + lineRange[1],
            });
        }

        let summaryOutput;
        if (this.props.summaryText) {
            const summaryTree = PerseusMarkdown.parseInline(
                this.props.summaryText,
            );
            summaryOutput = (
                <span aria-hidden={true}>
                    {" "}
                    {/* curly quotes */}
                    (&ldquo;
                    {PerseusMarkdown.basicOutput(summaryTree)}
                    &rdquo;)
                </span>
            );
        } else {
            summaryOutput = null;
        }

        return (
            <span>
                {lineRangeOutput}
                {summaryOutput}
                {lineRange && (
                    <div className="perseus-sr-only">{this.state.content}</div>
                )}
            </span>
        );
    }

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    componentDidMount() {
        // TODO(scottgrant): This is a hack to remove the deprecated call to
        // this.isMounted() but is still considered an anti-pattern.
        this._isMounted = true;

        this._deferredUpdateRange();

        this._throttledUpdateRange = _.throttle(this._deferredUpdateRange, 500);
        window.addEventListener("resize", this._throttledUpdateRange);
    }

    componentDidUpdate() {
        this._deferredUpdateRange();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this._throttledUpdateRange);

        this._isMounted = false;
    }

    _deferredUpdateRange: () => void = () => {
        _.defer(this._updateRange);
    };

    _updateRange: () => void = () => {
        // Note(TB): findWidgets runs findInternal and findExternal;
        // findExternal runs findInternal for the renderers involved;
        // findInternal returns type $ReadOnlyArray<?Widget>
        const passage: ?PassageWidgetExport.widget = this.props.findWidgets(
            "passage " + this.props.passageNumber,
        )[0];

        let refInfo: ?Reference = null;
        if (passage) {
            refInfo = passage.getReference(this.props.referenceNumber);
        }

        if (this._isMounted) {
            if (refInfo) {
                this.setState({
                    lineRange: [refInfo.startLine, refInfo.endLine],
                    content: refInfo.content,
                });
            } else {
                this.setState({
                    lineRange: null,
                    content: null,
                });
            }
        }
    };

    simpleValidate: (Rubric) => PerseusScore = (rubric) => {
        return PassageRef.validate(this.getUserInput(), rubric);
    };

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }
}

export default ({
    name: "passage-ref",
    displayName: "PassageRef (SAT only)",
    defaultAlignment: "inline",
    widget: PassageRef,
    transform: (
        widgetOptions: PerseusPassageRefWidgetOptions,
    ): RenderProps => ({
        passageNumber: widgetOptions.passageNumber,
        referenceNumber: widgetOptions.referenceNumber,
        summaryText: widgetOptions.summaryText,
    }),
    version: {major: 0, minor: 1},
}: WidgetExports<typeof PassageRef>);
