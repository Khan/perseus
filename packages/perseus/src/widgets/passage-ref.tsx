/* eslint-disable react/sort-comp */
import * as i18n from "@khanacademy/wonder-blocks-i18n";
import * as React from "react";
import _ from "underscore";

import * as Changeable from "../mixins/changeable";
import {removeDenylistProps} from "../mixins/widget-prop-denylist";
import PerseusMarkdown from "../perseus-markdown";

import type {Passage, Reference} from "./passage";
import type {PerseusPassageRefWidgetOptions} from "../perseus-types";
import type {
    ChangeFn,
    PerseusScore,
    WidgetExports,
    WidgetProps,
} from "../types";
import type {LinterContextProps} from "@khanacademy/perseus-linter";

const EN_DASH = "\u2013";

type UserInput = RenderProps & {
    static: boolean | null | undefined;
    reviewModeRubric: Rubric;
    linterContext: LinterContextProps;
    isLastUsedWidget: boolean;
    alignment: string | null | undefined;
};

type RenderProps = {
    passageNumber: PerseusPassageRefWidgetOptions["passageNumber"];
    referenceNumber: PerseusPassageRefWidgetOptions["referenceNumber"];
    summaryText: PerseusPassageRefWidgetOptions["summaryText"];
};

type Rubric = PerseusPassageRefWidgetOptions;

type Props = WidgetProps<RenderProps, Rubric>;

type DefaultProps = {
    passageNumber: Props["passageNumber"];
    referenceNumber: Props["referenceNumber"];
    summaryText: Props["summaryText"];
};

type State = {
    lineRange: [number, number] | null | undefined;
    content: string | null | undefined;
};

class PassageRef extends React.Component<Props, State> {
    displayName = "PassageRef";
    // @ts-expect-error - TS2564 - Property '_isMounted' has no initializer and is not definitely assigned in the constructor.
    _isMounted: boolean;
    // @ts-expect-error - TS2564 - Property '_throttledUpdateRange' has no initializer and is not definitely assigned in the constructor.
    _throttledUpdateRange: () => void;

    static defaultProps: DefaultProps = {
        passageNumber: 1,
        referenceNumber: 1,
        summaryText: "",
    };

    state: State = {
        lineRange: null,
        content: null,
    };

    shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
        return (
            !_.isEqual(this.props, nextProps) ||
            !_.isEqual(this.state, nextState)
        );
    }

    // TODO(tamarab): getUserInput needs to be updated to only return
    // props input by the user. Currently returns all the widget's props.
    getUserInput: () => UserInput = () => {
        return removeDenylistProps(this.props);
    };

    render(): React.ReactNode {
        const lineRange = this.state.lineRange;
        let lineRangeOutput;
        if (!lineRange) {
            lineRangeOutput = i18n.$_("lines %(lineRange)s", {
                lineRange: `?${EN_DASH}?`,
            });
        } else if (lineRange[0] === lineRange[1]) {
            lineRangeOutput = i18n.$_("line %(lineNumber)s", {
                lineNumber: lineRange[0],
            });
        } else {
            lineRangeOutput = i18n.$_("lines %(lineRange)s", {
                lineRange: lineRange[0] + EN_DASH + lineRange[1],
            });
        }

        let summaryOutput;
        if (this.props.summaryText) {
            // @ts-expect-error - TS2554 - Expected 2 arguments, but got 1.
            const summaryTree = PerseusMarkdown.parseInline(
                this.props.summaryText,
            );
            summaryOutput = (
                <span aria-hidden={true}>
                    {" "}
                    {/* curly quotes */}
                    (&ldquo;
                    {PerseusMarkdown.basicOutput(summaryTree)}
                    &rdquo;)
                </span>
            );
        } else {
            summaryOutput = null;
        }

        return (
            <span>
                {lineRangeOutput}
                {summaryOutput}
                {lineRange && (
                    <div className="perseus-sr-only">{this.state.content}</div>
                )}
            </span>
        );
    }

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    componentDidMount() {
        // TODO(scottgrant): This is a hack to remove the deprecated call to
        // this.isMounted() but is still considered an anti-pattern.
        this._isMounted = true;

        this._deferredUpdateRange();

        this._throttledUpdateRange = _.throttle(this._deferredUpdateRange, 500);
        window.addEventListener("resize", this._throttledUpdateRange);
    }

    componentDidUpdate() {
        this._deferredUpdateRange();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this._throttledUpdateRange);

        this._isMounted = false;
    }

    _deferredUpdateRange: () => void = () => {
        _.defer(this._updateRange);
    };

    _updateRange: () => void = () => {
        // Note(TB): findWidgets runs findInternal and findExternal;
        // findExternal runs findInternal for the renderers involved;
        // findInternal returns type $ReadOnlyArray<?Widget>
        const passage: Passage | null | undefined = this.props.findWidgets(
            "passage " + this.props.passageNumber,
        )[0];

        let refInfo: Reference | null | undefined = null;
        if (passage) {
            refInfo = passage.getReference(this.props.referenceNumber);
        }

        if (this._isMounted) {
            if (refInfo) {
                this.setState({
                    lineRange: [refInfo.startLine, refInfo.endLine],
                    content: refInfo.content,
                });
            } else {
                this.setState({
                    lineRange: null,
                    content: null,
                });
            }
        }
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return PassageRef.validate(this.getUserInput(), rubric);
    };

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }
}

export default {
    name: "passage-ref",
    displayName: "PassageRef (SAT only)",
    defaultAlignment: "inline",
    widget: PassageRef,
    transform: (
        widgetOptions: PerseusPassageRefWidgetOptions,
    ): RenderProps => ({
        passageNumber: widgetOptions.passageNumber,
        referenceNumber: widgetOptions.referenceNumber,
        summaryText: widgetOptions.summaryText,
    }),
    version: {major: 0, minor: 1},
} as WidgetExports<typeof PassageRef>;
