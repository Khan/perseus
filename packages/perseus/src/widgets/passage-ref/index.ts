export {default} from "./passage-ref";
