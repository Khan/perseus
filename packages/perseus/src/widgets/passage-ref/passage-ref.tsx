import * as React from "react";
import _ from "underscore";

import {PerseusI18nContext} from "../../components/i18n-context";
import * as Changeable from "../../mixins/changeable";
import PerseusMarkdown from "../../perseus-markdown";
import noopValidator from "../__shared__/noop-validator";
import {isPassageWidget} from "../passage/utils";

import type {PerseusPassageRefWidgetOptions} from "../../perseus-types";
import type {ChangeFn, Widget, WidgetExports, WidgetProps} from "../../types";

const EN_DASH = "\u2013";

type RenderProps = {
    passageNumber: PerseusPassageRefWidgetOptions["passageNumber"];
    referenceNumber: PerseusPassageRefWidgetOptions["referenceNumber"];
    summaryText: PerseusPassageRefWidgetOptions["summaryText"];
};

type Props = WidgetProps<RenderProps>;

type DefaultProps = {
    passageNumber: Props["passageNumber"];
    referenceNumber: Props["referenceNumber"];
    summaryText: Props["summaryText"];
};

type State = {
    lineRange: [number, number] | null | undefined;
    content: string | null | undefined;
};

class PassageRef extends React.Component<Props, State> implements Widget {
    static contextType = PerseusI18nContext;
    declare context: React.ContextType<typeof PerseusI18nContext>;

    displayName = "PassageRef";
    // @ts-expect-error - TS2564 - Property '_isMounted' has no initializer and is not definitely assigned in the constructor.
    _isMounted: boolean;
    // @ts-expect-error - TS2564 - Property '_throttledUpdateRange' has no initializer and is not definitely assigned in the constructor.
    _throttledUpdateRange: () => void;

    static defaultProps: DefaultProps = {
        passageNumber: 1,
        referenceNumber: 1,
        summaryText: "",
    };

    // this just helps with TS weak typing when a Widget
    // doesn't implement any Widget methods
    isWidget = true as const;

    state: State = {
        lineRange: null,
        content: null,
    };

    componentDidMount() {
        // TODO(scottgrant): This is a hack to remove the deprecated call to
        // this.isMounted() but is still considered an anti-pattern.
        this._isMounted = true;

        this._deferredUpdateRange();

        this._throttledUpdateRange = _.throttle(this._deferredUpdateRange, 500);
        window.addEventListener("resize", this._throttledUpdateRange);
    }

    shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
        return (
            !_.isEqual(this.props, nextProps) ||
            !_.isEqual(this.state, nextState)
        );
    }

    componentDidUpdate() {
        this._deferredUpdateRange();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this._throttledUpdateRange);

        this._isMounted = false;
    }

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    _deferredUpdateRange: () => void = () => {
        _.defer(this._updateRange);
    };

    _updateRange: () => void = () => {
        const passage = this.props
            .findWidgets("passage " + this.props.passageNumber)
            .filter(isPassageWidget)[0];

        const refInfo = passage?.getReference(this.props.referenceNumber);

        if (this._isMounted) {
            if (refInfo) {
                this.setState({
                    lineRange: [refInfo.startLine, refInfo.endLine],
                    content: refInfo.content,
                });
            } else {
                this.setState({
                    lineRange: null,
                    content: null,
                });
            }
        }
    };

    render(): React.ReactNode {
        const {strings} = this.context;
        const lineRange = this.state.lineRange;
        let lineRangeOutput;
        if (!lineRange) {
            lineRangeOutput = strings.lineRange({
                lineRange: `?${EN_DASH}?`,
            });
        } else if (lineRange[0] === lineRange[1]) {
            lineRangeOutput = strings.lineNumber({
                lineNumber: String(lineRange[0]),
            });
        } else {
            lineRangeOutput = strings.lineRange({
                lineRange: lineRange[0] + EN_DASH + lineRange[1],
            });
        }

        let summaryOutput;
        if (this.props.summaryText) {
            // @ts-expect-error - TS2554 - Expected 2 arguments, but got 1.
            const summaryTree = PerseusMarkdown.parseInline(
                this.props.summaryText,
            );
            summaryOutput = (
                <span aria-hidden={true}>
                    {" "}
                    {/* curly quotes */}
                    (&ldquo;
                    {PerseusMarkdown.basicOutput(summaryTree)}
                    &rdquo;)
                </span>
            );
        } else {
            summaryOutput = null;
        }

        return (
            <span>
                {lineRangeOutput}
                {summaryOutput}
                {lineRange && (
                    <div className="perseus-sr-only">{this.state.content}</div>
                )}
            </span>
        );
    }
}

export default {
    name: "passage-ref",
    displayName: "PassageRef (SAT only)",
    hidden: true,
    defaultAlignment: "inline",
    widget: PassageRef,
    transform: (
        widgetOptions: PerseusPassageRefWidgetOptions,
    ): RenderProps => ({
        passageNumber: widgetOptions.passageNumber,
        referenceNumber: widgetOptions.referenceNumber,
        summaryText: widgetOptions.summaryText,
    }),
    version: {major: 0, minor: 1},
    // TODO: things that aren't interactive shouldn't need validators
    validator: () => noopValidator(),
} as WidgetExports<typeof PassageRef>;
