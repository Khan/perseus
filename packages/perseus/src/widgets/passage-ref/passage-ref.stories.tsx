import {generateTestPerseusItem} from "@khanacademy/perseus-core";

import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";

import {question1, question2} from "./passage-ref.testdata";

import type {Meta, StoryObj} from "@storybook/react";

const meta: Meta = {
    title: "Perseus/Widgets/PassageRef",
    component: ServerItemRendererWithDebugUI,
};
export default meta;

type Story = StoryObj<typeof ServerItemRendererWithDebugUI>;

export const ShortPassage: Story = {
    args: {
        item: generateTestPerseusItem({question: question1}),
    },
};

export const LongPassage: Story = {
    args: {
        item: generateTestPerseusItem({question: question2}),
    },
};
