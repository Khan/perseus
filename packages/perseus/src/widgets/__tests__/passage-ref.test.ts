import {screen} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {question1} from "../__testdata__/passage-ref.testdata";
import PassageExport from "../passage";

import {renderQuestion} from "./renderQuestion";

const mockReference = (
    mock: null | {
        startLine: number;
        endLine: number;
        content: string;
    },
) => {
    jest.spyOn(
        PassageExport.widget.prototype,
        "getReference",
    ).mockImplementation(() => {
        return mock;
    });
};

describe("passage-ref widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should render with invalid reference", () => {
        // Arrange
        mockReference(null);

        // Act
        const {container} = renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should render unkown line numbers with invalid reference", () => {
        // Arrange
        mockReference(null);

        // Act
        renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(screen.getByText("lines ?–?", {exact: false})).toBeVisible();
    });

    it("should render with single-line reference", () => {
        // Arrange
        mockReference({
            startLine: 4,
            endLine: 4,
            content: "mocked reference content",
        });

        // Act
        const {container} = renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should render correct line number with single-line reference", () => {
        // Arrange
        mockReference({
            startLine: 4,
            endLine: 4,
            content: "mocked reference content",
        });

        // Act
        renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(screen.getByText("line 4", {exact: false})).toBeVisible();
    });

    it("should render with multi-line reference", () => {
        // Arrange
        mockReference({
            startLine: 4,
            endLine: 6,
            content: "mocked reference content",
        });

        // Act
        const {container} = renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should render correct line numbers with multi-line reference", () => {
        // Arrange
        mockReference({
            startLine: 4,
            endLine: 6,
            content: "mocked reference content",
        });

        // Act
        renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(screen.getByText("lines 4–6", {exact: false})).toBeVisible();
    });

    it("should render with summary text", () => {
        // Arrange
        mockReference({
            startLine: 4,
            endLine: 4,
            content: "mocked reference content",
        });

        // Act
        renderQuestion({
            ...question1,
            widgets: {
                ...question1.widgets,
                "passage-ref 1": {
                    options: {
                        passageNumber: 1,
                        referenceNumber: 1,
                        summaryText: "A **summary** of the _reference_",
                    },
                    type: "passage-ref",
                    version: {
                        major: 0,
                        minor: 1,
                    },
                },
            },
        });
        jest.runOnlyPendingTimers();

        // Assert
        expect(
            screen.getByText("of the", {
                exact: false,
            }),
        ).toHaveTextContent("A summary of the reference");
    });

    it("should not be answerable", () => {
        // Arrange

        // Act
        const {renderer} = renderQuestion(question1);
        jest.runOnlyPendingTimers();

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
