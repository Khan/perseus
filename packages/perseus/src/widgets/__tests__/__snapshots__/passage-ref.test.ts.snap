// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passage-ref widget should render with invalid reference 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div>
            <div
              class="perseus-widget-passage-container"
            >
              <div
                class="perseus-widget-passage-instructions"
              >
                

              </div>
              <div
                class="perseus-widget-passage"
              >
                <h3
                  class="perseus-sr-only"
                >
                  Beginning of reading passage.
                </h3>
                <div
                  class="passage-text"
                >
                  <div
                    class="container_6htn2u"
                  >
                    <div>
                      <div>
                        <div>
                          <div />
                        </div>
                      </div>
                    </div>
                    <div
                      class="content_gor68n"
                    >
                      <div>
                        <div>
                          <div
                            aria-hidden="true"
                            class="measurer_225hir"
                          >
                            <div
                              class="paragraph"
                            >
                              Line Height Measurement
                            </div>
                          </div>
                          <div
                            class="paragraph"
                          >
                            <span>
                              Our electronic devices are keeping us awake—not just because we spend time using them when we should be sleeping
                            </span>
                            <span>
                              , but also because they emit a blue light that keeps our brains awake
                            </span>
                            <span>
                              . One study showed that people who used screens that emitted blue light were unable to produce sufficient melatonin
                            </span>
                            <span>
                              , a hormone that 
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                              promotes
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                               sleep
                            </span>
                            <span>
                              . In order to enjoy a more restful night
                            </span>
                            <span>
                              , experts recommend shutting off screens at least two hours before bed
                            </span>
                            <span>
                              . 
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="perseus-sr-only"
                >
                  End of reading passage.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        We can see the word “promotes” is used in 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <span>
            lines 
            ?–?
            
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`passage-ref widget should render with multi-line reference 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div>
            <div
              class="perseus-widget-passage-container"
            >
              <div
                class="perseus-widget-passage-instructions"
              >
                

              </div>
              <div
                class="perseus-widget-passage"
              >
                <h3
                  class="perseus-sr-only"
                >
                  Beginning of reading passage.
                </h3>
                <div
                  class="passage-text"
                >
                  <div
                    class="container_6htn2u"
                  >
                    <div>
                      <div>
                        <div>
                          <div />
                        </div>
                      </div>
                    </div>
                    <div
                      class="content_gor68n"
                    >
                      <div>
                        <div>
                          <div
                            aria-hidden="true"
                            class="measurer_225hir"
                          >
                            <div
                              class="paragraph"
                            >
                              Line Height Measurement
                            </div>
                          </div>
                          <div
                            class="paragraph"
                          >
                            <span>
                              Our electronic devices are keeping us awake—not just because we spend time using them when we should be sleeping
                            </span>
                            <span>
                              , but also because they emit a blue light that keeps our brains awake
                            </span>
                            <span>
                              . One study showed that people who used screens that emitted blue light were unable to produce sufficient melatonin
                            </span>
                            <span>
                              , a hormone that 
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                              promotes
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                               sleep
                            </span>
                            <span>
                              . In order to enjoy a more restful night
                            </span>
                            <span>
                              , experts recommend shutting off screens at least two hours before bed
                            </span>
                            <span>
                              . 
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="perseus-sr-only"
                >
                  End of reading passage.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        We can see the word “promotes” is used in 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <span>
            lines 
            4–6
            
            <div
              class="perseus-sr-only"
            >
              mocked reference content
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`passage-ref widget should render with single-line reference 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div>
            <div
              class="perseus-widget-passage-container"
            >
              <div
                class="perseus-widget-passage-instructions"
              >
                

              </div>
              <div
                class="perseus-widget-passage"
              >
                <h3
                  class="perseus-sr-only"
                >
                  Beginning of reading passage.
                </h3>
                <div
                  class="passage-text"
                >
                  <div
                    class="container_6htn2u"
                  >
                    <div>
                      <div>
                        <div>
                          <div />
                        </div>
                      </div>
                    </div>
                    <div
                      class="content_gor68n"
                    >
                      <div>
                        <div>
                          <div
                            aria-hidden="true"
                            class="measurer_225hir"
                          >
                            <div
                              class="paragraph"
                            >
                              Line Height Measurement
                            </div>
                          </div>
                          <div
                            class="paragraph"
                          >
                            <span>
                              Our electronic devices are keeping us awake—not just because we spend time using them when we should be sleeping
                            </span>
                            <span>
                              , but also because they emit a blue light that keeps our brains awake
                            </span>
                            <span>
                              . One study showed that people who used screens that emitted blue light were unable to produce sufficient melatonin
                            </span>
                            <span>
                              , a hormone that 
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                              promotes
                            </span>
                            <span
                              style="display: inline-block; width: 0px; visibility: hidden;"
                            >
                              _
                            </span>
                            <span>
                               sleep
                            </span>
                            <span>
                              . In order to enjoy a more restful night
                            </span>
                            <span>
                              , experts recommend shutting off screens at least two hours before bed
                            </span>
                            <span>
                              . 
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="perseus-sr-only"
                >
                  End of reading passage.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        We can see the word “promotes” is used in 
        <div
          class="perseus-widget-container widget-nohighlight widget-inline"
        >
          <span>
            line 
            4
            
            <div
              class="perseus-sr-only"
            >
              mocked reference content
            </div>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
